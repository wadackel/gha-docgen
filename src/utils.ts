import * as fs from 'node:fs/promises';
import type { ZodError } from 'zod';
import { fromZodError } from 'zod-validation-error';

/**
 * Filesystem
 */
export const read = async (filepath: string): Promise<[Error, null] | [null, string]> => {
  try {
    const data = await fs.readFile(filepath, 'utf8');
    return [null, data];
  } catch (e) {
    return [e instanceof Error ? e : new Error(String(e)), null];
  }
};

export const write = async (filepath: string, data: string): Promise<Error | null> => {
  try {
    await fs.writeFile(filepath, data, 'utf8');
    return null;
  } catch (e) {
    return e instanceof Error ? e : new Error(String(e));
  }
};

/**
 * Error handling
 */
export const generateZodErrorMessage = (error: ZodError): string => {
  return fromZodError(error, {
    prefix: undefined,
    issueSeparator: ' | ',
    unionSeparator: ' OR ',
  }).toString();
};

/**
 * Strings
 */
const NEWLINE_REGEX = /\r\n|\r|\n/g;

export const nl2space = (input: string): string => {
  return input.replace(NEWLINE_REGEX, ' ');
};

export const nl2md = (input: string): string => {
  return input.replace(NEWLINE_REGEX, '  \n');
};

export const nl2br = (input: string): string => {
  return input.replace(NEWLINE_REGEX, '<br />');
};
