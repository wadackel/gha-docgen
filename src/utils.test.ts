import { expect, test } from 'vitest';
import { z } from 'zod';
import { generateZodErrorMessage, nl2br, nl2md, nl2space } from './utils';

/**
 * Error handling
 */
test('generateZodErrorMessage', () => {
  const schema = z.object({
    foo: z.string(),
    bar: z.object({
      baz: z.number(),
    }),
  });

  const res = schema.safeParse({
    foo: 0,
    bar: {
      baz: 'str',
    },
  });

  if (res.success) {
    throw new Error();
  }

  expect(generateZodErrorMessage(res.error.issues)).toMatchSnapshot();
});

/**
 * Strings
 */
test.each([
  // prettier-ignore
  ['line1', 'line1'],
  ['line2<br>line2', 'line2<br>line2'],
  ['line1\nline2', 'line1 line2'],
])('nl2space - %s', (input, expected) => {
  expect(nl2space(input)).toBe(expected);
});

test.each([
  // prettier-ignore
  ['line1', 'line1'],
  ['line2<br>line2', 'line2<br>line2'],
  ['line1\nline2', 'line1  \nline2'],
])('nl2md - %s', (input, expected) => {
  expect(nl2md(input)).toBe(expected);
});

test.each([
  // prettier-ignore
  ['line1', 'line1'],
  ['line2<br>line2', 'line2<br>line2'],
  ['line1\nline2', 'line1<br />line2'],
])('nl2br - %s', (input, expected) => {
  expect(nl2br(input)).toBe(expected);
});
