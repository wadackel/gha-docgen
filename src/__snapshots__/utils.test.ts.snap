// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateZodErrorMessage 1`] = `"Invalid input: expected string, received number at "foo" | Invalid input: expected number, received string at "bar.baz""`;
